doc/design.md

```diff
@@ -0,0 +1,922 @@
+- Data variables have a name in addition to their tag.
+  *Example: We can have a data variable holding neutron counts in `Data::Value` fom the "sample" run and a data variable holding counts for the "can" run in the *same* dataset.*
+
+The name of data variables is furthermore used to imply a grouping of variables.
+*Example: In a dataset containing data variables `Data::Value` and `Data::Variance` a matching variable name implies that the uncertainties stored in `Data::Variance` are associated with the data stored in `Data::Value`.*
+
+The data in `Variable` is held by a type-erased handle, i.e., a `Variable` can hold data of any type.
+The data handle also implements a copy-on-write mechanism, which makes `Variable` and `Dataset` cheap to copy and saves memory if only some variables in a dataset are modified.
+
+
+### <a name="overview-operations"></a>Operations
+
+Alongside the new `Dataset` type a series of basic and common operations will be provided.
+A mostly but not entirely complete list of operations to be supported is:
+1. Arithmetic operations: unary `-`, binary `+`, `-`, `*`, `/`
+1. `transpose`
+1. `slice` and `dice`
+1. `concatenate` and `merge`
+1. `sort`
+1. `rebin`
+1. `convert` (unit and/or dimension)
+1. `filter`
+1. Boolean/logical operations for mask operations
+1. `integrate`
+1. Basic statistics operations such as min, max, mean, and standard deviation
+1. Serialization and deserialization, mainly for saving to and loading from disk
+   *Question: Should disk I/O live in a separate module/library?*
+
+
+### <a name="overview-relation-to-existing-workspace-types"></a>Relation to existing workspace types
+
+From the above description it may not be clear how `Dataset` is used to replace existing workspace types or components so we take a break to give some examples.
+For clarity, in the following we are explicitly giving the current Mantid namespace names such as `API`, `DataObjects`, `Geometry`, and `Algorithms` when referring to types from the existing implementation.
+
+##### DataObjects::TableWorkspace
+
+A dataset where all variables are one-dimensional is a table.
+- Each variable is a column.
+- The variable name is the column label.
+- `Coord::RowLabel` can provide a row label.
+- `concatenate` is used to append one table to another.
+- `merge` combines columns from two tables into one.
+- `slice`, `dice`, or `filter` extract a new table with one or more rows.
+- `sort` sorts the table.
+- Coordinate and name matching enables mathematical operations on individual table columns.
+
+##### DataObjects::MDHistoWorkspace
+
+A dataset with multi-dimensional variables `Data::Value` and `Data::Variance` in combination with a corresponding number of one--dimensional coordinate variables is a close equivalent to the current `DataObjects::MDHistoWorkspace`.
+A typical example would be three-dimensional `Data::Value` and `Data::Variance` alongside one-dimensional `Coord:Qx`, `Coord:Qy`, and `Coord::Qz`.
```

Comment:
See below regarding setting precision at compile time. Having it completely flexibly would lead to a nightmarish template mess (in particular relating to branching into the right template variant at run-time, due to the type erasure) if we require all algorithms to support all possible combinations of precisions.